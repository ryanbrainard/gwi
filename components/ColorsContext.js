import React from "react";

export default (ColorsContext = React.createContext());

export function provideColors(colors, Component) {
  return props => {
    return (
      <ColorsContext.Provider value={colors}>
        <Component {...props} />
      </ColorsContext.Provider>
    );
  };
}

export function withColors(Component) {
  return props => {
    return (
      <ColorsContext.Consumer>
        {colors => <Component {...props} colors={colors} />}
      </ColorsContext.Consumer>
    );
  };
}
