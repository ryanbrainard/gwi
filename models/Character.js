import VoicePlayerMRU from "./VoicePlayerMRU";

export default class Character {
  static _chars = {};

  static register(char) {
    this._chars[char.name] = char;
    return char;
  }

  static find(name) {
    return this._chars[name];
  }

  constructor(name, group, voices) {
    this._name = name;
    this._group = group;
    this._voices = voices;

    this.player = new VoicePlayerMRU(this._voices);
  }

  get key() {
    return this.name;
  }

  get name() {
    return this._name;
  }

  get group() {
    return this._group.map(c => Character.find(c));
  }

  async preloadPlay() {
    await this.player.load();
  }

  async play() {
    await this.player.play();
  }
}
