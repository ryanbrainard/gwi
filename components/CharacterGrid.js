import PropTypes from "prop-types";
import React from "react";
import { ActivityIndicator, FlatList, StyleSheet, View } from "react-native";
import config from "../config";
import Character from "../models/Character";
import CharacterTile from "./CharacterTile";

export default class CharacterGrid extends React.Component {
  static propTypes = {
    chars: PropTypes.arrayOf(PropTypes.instanceOf(Character)).isRequired,
    color: PropTypes.string,
    parentLayout: PropTypes.shape({
      width: PropTypes.number,
      height: PropTypes.number
    })
  };

  render() {
    const { chars, color, parentLayout } = this.props;

    if (!parentLayout) {
      return (
        <View style={styles.activityIndicatorContainer}>
          <ActivityIndicator size="large" color={color} />
        </View>
      );
    }

    const isPortrait = parentLayout.height > parentLayout.width;
    const cols = isPortrait ? 3 : 5;
    const size = (parentLayout.width / cols) * 0.9;

    return (
      <View style={styles.container}>
        <FlatList
          key={cols /* force refresh, per recommendation */}
          data={chars}
          renderItem={({ item }) => (
            <CharacterTile char={item} size={size} color={color} />
          )}
          numColumns={cols}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: config.colors.background,
    alignItems: "center"
  },
  activityIndicatorContainer: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center"
  }
});
