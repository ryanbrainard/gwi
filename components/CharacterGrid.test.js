import React from "react";
import Character from "../models/Character";
import CharacterGrid from "./CharacterGrid";
import renderer from "react-test-renderer";
import { provideColors } from "./ColorsContext";

test("renders correctly", () => {
  const _CharacterGrid = provideColors({}, CharacterGrid);
  const tree = renderer
    .create(
      <_CharacterGrid
        chars={[new Character("A"), new Character("B")]}
        parentLayout={{ width: 100, height: 200 }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
