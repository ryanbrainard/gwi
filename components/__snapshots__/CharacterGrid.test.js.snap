// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Character {
          "_group": undefined,
          "_name": "A",
          "_voices": undefined,
          "player": VoicePlayerMRU {
            "playAfterLoad": false,
            "playback": Sound {
              "_coalesceStatusUpdatesInMillis": 100,
              "_errorCallback": [Function],
              "_eventEmitter": NativeEventEmitter {
                "_nativeModule": Object {
                  "addListener": [MockFunction],
                  "getAudioRecordingStatus": [MockFunction],
                  "getStatusForSound": [MockFunction],
                  "getStatusForVideo": [MockFunction],
                  "loadForSound": [MockFunction],
                  "loadForVideo": [MockFunction],
                  "pauseAudioRecording": [MockFunction],
                  "prepareAudioRecorder": [MockFunction],
                  "removeListeners": [MockFunction],
                  "replaySound": [MockFunction],
                  "replayVideo": [MockFunction],
                  "setAudioIsEnabled": [MockFunction],
                  "setAudioMode": [MockFunction],
                  "setErrorCallbackForSound": [MockFunction],
                  "setStatusForSound": [MockFunction],
                  "setStatusForVideo": [MockFunction],
                  "startAudioRecording": [MockFunction],
                  "stopAudioRecording": [MockFunction],
                  "unloadAudioRecorder": [MockFunction],
                  "unloadForSound": [MockFunction],
                  "unloadForVideo": [MockFunction],
                },
                "_subscriber": EventSubscriptionVendor {
                  "_currentSubscription": null,
                  "_subscriptionsForType": Object {
                    "didUpdateDimensions": Array [
                      EmitterSubscription {
                        "context": undefined,
                        "emitter": RCTDeviceEventEmitter {
                          "_subscriber": [Circular],
                          "sharedSubscriber": [Circular],
                        },
                        "eventType": "didUpdateDimensions",
                        "key": 0,
                        "listener": [Function],
                        "subscriber": [Circular],
                      },
                    ],
                  },
                },
              },
              "_internalStatusUpdateCallback": [Function],
              "_key": -1,
              "_lastStatusUpdate": null,
              "_lastStatusUpdateTime": null,
              "_loaded": false,
              "_loading": false,
              "_onPlaybackStatusUpdate": null,
              "_subscriptions": Array [],
              "getStatusAsync": [Function],
            },
            "voices": undefined,
          },
        },
        Character {
          "_group": undefined,
          "_name": "B",
          "_voices": undefined,
          "player": VoicePlayerMRU {
            "playAfterLoad": false,
            "playback": Sound {
              "_coalesceStatusUpdatesInMillis": 100,
              "_errorCallback": [Function],
              "_eventEmitter": NativeEventEmitter {
                "_nativeModule": Object {
                  "addListener": [MockFunction],
                  "getAudioRecordingStatus": [MockFunction],
                  "getStatusForSound": [MockFunction],
                  "getStatusForVideo": [MockFunction],
                  "loadForSound": [MockFunction],
                  "loadForVideo": [MockFunction],
                  "pauseAudioRecording": [MockFunction],
                  "prepareAudioRecorder": [MockFunction],
                  "removeListeners": [MockFunction],
                  "replaySound": [MockFunction],
                  "replayVideo": [MockFunction],
                  "setAudioIsEnabled": [MockFunction],
                  "setAudioMode": [MockFunction],
                  "setErrorCallbackForSound": [MockFunction],
                  "setStatusForSound": [MockFunction],
                  "setStatusForVideo": [MockFunction],
                  "startAudioRecording": [MockFunction],
                  "stopAudioRecording": [MockFunction],
                  "unloadAudioRecorder": [MockFunction],
                  "unloadForSound": [MockFunction],
                  "unloadForVideo": [MockFunction],
                },
                "_subscriber": EventSubscriptionVendor {
                  "_currentSubscription": null,
                  "_subscriptionsForType": Object {
                    "didUpdateDimensions": Array [
                      EmitterSubscription {
                        "context": undefined,
                        "emitter": RCTDeviceEventEmitter {
                          "_subscriber": [Circular],
                          "sharedSubscriber": [Circular],
                        },
                        "eventType": "didUpdateDimensions",
                        "key": 0,
                        "listener": [Function],
                        "subscriber": [Circular],
                      },
                    ],
                  },
                },
              },
              "_internalStatusUpdateCallback": [Function],
              "_key": -1,
              "_lastStatusUpdate": null,
              "_lastStatusUpdateTime": null,
              "_loaded": false,
              "_loading": false,
              "_onPlaybackStatusUpdate": null,
              "_subscriptions": Array [],
              "getStatusAsync": [Function],
            },
            "voices": undefined,
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    invertStickyHeaders={undefined}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "borderColor": "#FFFFFF",
                "borderWidth": 2,
              }
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": undefined,
                  "height": 30.000000000000004,
                  "justifyContent": "center",
                  "opacity": 1,
                  "width": 30.000000000000004,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#000000",
                    "fontSize": 24,
                    "fontWeight": "bold",
                  }
                }
              >
                A
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "borderColor": "#FFFFFF",
                "borderWidth": 2,
              }
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": undefined,
                  "height": 30.000000000000004,
                  "justifyContent": "center",
                  "opacity": 1,
                  "width": 30.000000000000004,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#000000",
                    "fontSize": 24,
                    "fontWeight": "bold",
                  }
                }
              >
                B
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
